File: leetcode/go/273.integer-to-english-words.go
import (
    "strings"
)

var tens = [10]string {
    "", "Ten", "Twenty", "Thirty", "Forty", "Fifty", 
    "Sixty", "Seventy", "Eighty", "Ninety",
}

var numbers = [20]string {
    "", "One", "Two", "Three",
    "Four", "Five", "Six", "Seven",
    "Eight", "Nine", "Ten", "Eleven",
    "Twelve", "Thirteen", "Fourteen", "Fifteen",
    "Sixteen", "Seventeen", "Eighteen", "Nineteen",
}

func helper(num int) string {
    s := ""
    if num < 20 {
        s = numbers[num]
    } else if num < 100 {
        s = tens[num/10] + " " + numbers[num % 10]
    } else if num < 1000 {
        s = numbers[num/100] + " Hundred " + helper(num % 100)
    } else if num < 1000000 {
        s = helper(num/1000) + " Thousand " + helper(num % 1000)
    } else if num < 1000000000 {
        s = helper(num/1000000) + " Million " + helper(num % 1000000)
    } else {
        s = helper(num/1000000000) + " Billion " + helper(num % 1000000000)
    }
    s = strings.TrimSpace(s)
    return s
}

func numberToWords(num int) string {
    if num == 0 {
        return "Zero"
    }
    return helper(num)
}

File: leetcode/rust-leetcode/src/273.integer-to-english-words.rs
impl Solution {
    pub fn number_to_words(num: i32) -> String {
        if num == 0 {
            return String::from("Zero");
        }
        fn helper(num: usize) -> String {
            const tens: [&str; 10] = [
                "", "Ten", "Twenty", "Thirty", "Forty", "Fifty", 
                "Sixty", "Seventy", "Eighty", "Ninety",
            ];
            const numbers: [&str; 20] = [
                "", "One", "Two", "Three",
                "Four", "Five", "Six", "Seven",
                "Eight", "Nine", "Ten", "Eleven",
                "Twelve", "Thirteen", "Fourteen", "Fifteen",
                "Sixteen", "Seventeen", "Eighteen", "Nineteen",
            ];
            let mut s = String::new();
            match num {
                n if n < 20 => {
                    s.push_str(numbers[n]);
                },
                n if n < 100 => {
                    s.push_str(tens[n/10]);
                    s.push_str(" ");
                    s.push_str(numbers[n % 10]);
                },
                n if n < 1000 => {
                    s.push_str(numbers[n/100]);
                    s.push_str(" Hundred ");
                    s.push_str(&helper(n % 100));
                },
                n if n < 1000000 => {
                    s.push_str(&helper(n / 1000));
                    s.push_str(" Thousand ");
                    s.push_str(&helper(n % 1000));
                },
                n if n < 1000000000 => {
                    s.push_str(&helper(n / 1000000));
                    s.push_str(" Million ");
                    s.push_str(&helper(n % 1000000));
                },
                _ => {
                    s.push_str(&helper(num / 1000000000));
                    s.push_str(" Billion ");
                    s.push_str(&helper(num % 1000000000));
                }
            }
            s.trim().to_string()
        }
        helper(num as usize)
    }
}

